File: algorithms/22.py
# https://leetcode.com/problems/generate-parentheses


class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        result = []
        queue = [('', 0, n, n)]

        while len(queue) > 0:
            (s, w, lpq, rpq) = queue.pop(0)
            s += '('
            w += 1
            lpq -= 1
            if lpq == 0 or rpq == 0:
                s += ')' * rpq
                result.append(s)
            else:
                for r in range(rpq + 1):
                    if w - r >= 0:
                        queue.append((s + ')' * r, w - r, lpq, rpq - r))

        return result

File: algorithms/22.swift
// https://leetcode.com/problems/generate-parentheses

class Solution {
    func generateParenthesis(n: Int) -> [String] {
        var queue = [("", 0, n, n)]
        var result = [String]()

        while queue.count > 0 {
            var (s, w, lpq, rpq) = queue.removeFirst()
            s += "("
            w += 1
            lpq -= 1

            if lpq == 0 || rpq == 0 {
                let s_ = s + String(count: rpq, repeatedValue: ")" as Character)
                result.append(s_)
            } else {
                for r in 0...rpq {
                    if w-r >= 0 {
                        let s_ = s + String(count: r, repeatedValue: ")" as Character)
                        queue.append((s_, w-r, lpq, rpq-r))
                    }
                }
            }
        }

        return result
    }
}
